```{r echo=FALSE}
knitr::opts_chunk$set(cache=TRUE)
options(width=150)
```

```
Adapted from

################
# Differential Gene Expression Analysis on the Desktop
#
# Bio-IT World Asia conference 2012, SC1
# June 5, 10:05â€“10:35
#
# Charles Plessy
#################
```

# Preparation

## Install programs available in Debian

```
sudo apt-get install bedtools bwa paraclu python-pysam samtools sra-toolkit r-bioc-edger
```

In Debian _Wheezy_ (currently the _stable_ version), the package `python-pysam`
is only available from the `wheezy-backports` repository.  See
<http://backports.debian.org/Instructions/> for details.  The pacakge `paraclu`
is also only disponible in `unstable` at the moment.

## Install tagdust, and remove sequencer artefacts.

```
wget http://genome.gsc.riken.jp/osc/english/software/src/tagdust.tgz
tar xvf tagdust.tgz
(cd tagdust && make)
```

## Download and index the human genome.

```
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz
tar xvf chromFa.tar.gz
cat chr?.fa chr??.fa > hg19_male.fa
bwa index hg19_male.fa
```

## Install level1.py 

The following command will download some clustering scripts written in Python.
They require the `pysam` library, installed above with the `python-pysam`
package.

The script `level1.py` is then copied in the user's `bin` directory, to make
it available from the command line.

```
wget http://genome.gsc.riken.jp/plessy-20130430/PromoterPipeline_20130430.tar.gz
tar xvfz PromoterPipeline_20130430.tar.gz
cp PromoterPipeline/level1.py ~/bin
chmod ugo+x ~/bin/level1.py
```

# Download HeliScopeCAGE data from SRA.

```{r download_data, engine="sh"}
wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/DRX/DRX000/DRX000595/DRR000945/DRR000945.sra -O HeLa_1.sra
wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/DRX/DRX000/DRX000596/DRR000946/DRR000946.sra -O HeLa_2.sra
wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/DRX/DRX000/DRX000605/DRR000955/DRR000955.sra -O THP1_1.sra
wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/DRX/DRX000/DRX000606/DRR000956/DRR000956.sra -O THP1_2.sra
```

# Convert to FASTQ format

(this takes time).

```{r convert_to_fastq, dependson="download_data", engine="sh"}
fastq-dump *sra
```

# Filter artefacts with TagDust.

HeliScope artefacts are TAGC repreats.

```{r filter_artefacts, dependson="convert_to_fastq", engine="sh"}
cat > tagc.fa <<__END__
>tagc
tagctagctagctagctagctagctagctagctagctagc
__END__

for FQ in *fastq
do tagdust -fasta tagc.fa $FQ -o $(basename $FQ .fastq).dusted.fa -a $(basename $FQ .fastq).rejected.fa
done
```

# Align the tags, filter and sort the alignments.

```{r alignment, dependson="filter_artefacts", engine="sh"}
for FA in *.dusted.fa
do
  bwa aln  -t 8 hg19_male.fa -f $FA.sai $FA
  bwa samse     hg19_male.fa    $FA.sai $FA | samtools view -q20 -uS - | samtools sort - $(basename $FA .dusted.fa)
done
```

# Preapre an expression table.

```{r clustering, dependson="alignment", engine="sh"}
level1.py -o bioit.l1.osc.gz *bam
```

# In `R`

```{r load_data, dependson="clustering"}
l1 <- read.table('bioit.l1.osc.gz', row.names=1, head=TRUE)

THP1 <- c('raw.THP1_1', 'raw.THP1_2')
HeLa <- c('raw.HeLa_1', 'raw.HeLa_2')

with(l1, write.table(
  file="for_paraclu.txt",
  data.frame(chrom=chrom, strand=strand, start.0base=start.0base, total=rowSums(l1[,c(HeLa, THP1)])),
  col.names=FALSE, quote=FALSE, row.names=FALSE))
```

# Quit R and return to command line.

```
quit(save="yes")
```

# Cluster and filter

```{r paraclu, dependson="load_data", engine="bash"}
paraclu 0 for_paraclu.txt |
   paraclu-cut.sh -l 1000 > paraclu.txt
awk '{OFS="\t"}{print $1,$3,$4,$1":"$3"-"$4$2,$6,$2}' paraclu.txt |
   sort -k1,1 -k2,2n > paraclu.bed

# Create a correspondance table between TSS (single-nucleotide data) and clusters

zgrep -v \# bioit.l1.osc.gz | sed 1d | awk '{OFS="\t"}{print $2,$3,$4,$2":"$3"-"$4$5,"0",$5}' > bioit.bed
intersectBed -a bioit.bed -b paraclu.bed -wao -s | cut -f4,10 > l1-l2.txt
```

# In R again.

Count the number of tags per cluster.

```{r count_clusters_and_plot, dependson="paraclu", device="png"}
l1l2 <- read.table('l1-l2.txt', stringsAsFactors=FALSE)
colnames(l1l2) <- c("TSS", "Cluster")
l2 <- rowsum(l1[c(HeLa, THP1)], l1l2$Cluster)

# Pairwise plot of the samples, in log scale.

plot(log(l2))
```

# Plots the same to a file.

```
png('pairwise.png')
plot(log(l2))
dev.off()
```

# Digital expression analysis with edgeR

```{r DGE, dependson="count_clusters_and_plot"}
library("edgeR")

l2.dge <- DGEList(counts=l2, group=c('HeLa', 'HeLa', 'THP1', 'THP1'))
l2.dge <- calcNormFactors(l2.dge)
l2.dge <- estimateTagwiseDisp(l2.dge)
l2.com <- exactTest(l2.dge, dispersion="tagwise")
l2.detags <- rownames(l2.com$table[! decideTestsDGE(l2.com) ==0,])

topTags(l2.com)

# M-A plot

plotSmear(l2.com, de.tags=l2.detags)
```

# Plot the same into a file.

```
png('edgeR.png')
plotSmear(l2.com, de.tags=l2.detags)
dev.off()
```
