---
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

```{r echo=FALSE}
knitr::opts_chunk$set(cache=TRUE)
options(width=100)
```

# Digital expression comparison (using voom) between nanoCAGE libraries testing variant template-switching oligonucleotides

This does the same analysis as in the [main page](../analysis.md), but with `voom` from the [LIMMA](https://www.bioconductor.org/packages/limma) package.  It assumes that the main Rmarkdown file has already been [knitted](http://yihui.name/knitr/).


## Load the data.

```{r load_data}
library(magrittr)
library(oscR)

l2 <- read.osc("DRR014141.l2.osc.gz", drop.coord=T, drop.norm=T)

colnames(l2) %<>%
  sub('raw.DRR014141.',          '', .) %>%
  sub('_......_filtered_sorted', '', .)
```

## Organise the data.

The following commands defined convenient shortcuts to manipulates groups of
libraries. The presence of `nw_2` in the names indicate that strand-invasion
artifacts have been removed.

```{r define_aliases}
ddd <- c('ddd_1', 'ddd_2', 'ddd_3')
ddl <- c('ddl_1', 'ddl_2', 'ddl_3')
dll <- c('dll_1', 'dll_2', 'dll_3')
lll <- c('lll_1', 'lll_2', 'lll_3')
rrr <- c('rrr_1', 'rrr_2', 'rrr_3')
all <- c(rrr, lll, dll, ddl, ddd)

ddd_nw_2 <- c('ddd_1_nw_2', 'ddd_2_nw_2', 'ddd_3_nw_2')
ddl_nw_2 <- c('ddl_1_nw_2', 'ddl_2_nw_2', 'ddl_3_nw_2')
dll_nw_2 <- c('dll_1_nw_2', 'dll_2_nw_2', 'dll_3_nw_2')
lll_nw_2 <- c('lll_1_nw_2', 'lll_2_nw_2', 'lll_3_nw_2')
rrr_nw_2 <- c('rrr_1_nw_2', 'rrr_2_nw_2', 'rrr_3_nw_2')
all_nw_2 <- c(rrr_nw_2, lll_nw_2, dll_nw_2, ddl_nw_2, ddd_nw_2)
```

## Annotation of the results

```{r load_gene_symbols}
genesymbols <- read.table( "DRR014141.l2.genes"
                         , col.names=c("cluster","symbol"))
rownames(genesymbols) <- genesymbols$cluster
genesymbols$rmsk <- read.table( "DRR014141.l2.rmsk"
                              , col.names=c("cluster","rmsk"))[,'rmsk']
tail(genesymbols)
```

## Differential representation analysis

Statistical comparisons using [voom](https://www.bioconductor.org/packages/release/limma).

```{r voom, dev='png'}
library(limma)
library(edgeR)

# In alphabetical order
samples <- c(ddd, lll, rrr)
groups  <- c(rep('DNA',3), rep('LNA',3), rep('RNA', 3))

design <- model.matrix(~0+groups)
colnames(design) <- c("DNA", "LNA", "RNA")
design

cont.matrix <- makeContrasts( LNA_vs_RNA = LNA - RNA
                            , RNA_vs_DNA = RNA - DNA
                            , DNA_vs_LNA = DNA - LNA
                            , levels = design)

d <-  DGEList( counts = l2[samples]
             , group  = groups
             , remove.zeros = TRUE) %>%
        calcNormFactors %>%
        voom(design, plot = TRUE)

plotMDS(d)

d %<>% lmFit(design) %>%
  contrasts.fit(cont.matrix) %>%
  eBayes

d$genes <- genesymbols[rownames(d),]

topTable(d, n=30)
summary(decideTests(d))
decideTests(d) %>% vennDiagram
```

### LNA vs RNA

```{r compare_LNA_RNA}
n <- 1
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="LNA vs RNA")
topTreat(d,n)
```


### RNA vs DNA

```{r compare_RNA_DNA}
n <- 2
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="RNA vs DNA")
topTreat(d,n)
```


### DNA vs LNA

```{r compare_DNA_LNA}
n <- 3
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="DNA vs LNA")
topTreat(d,n)
```

## Differential representation analysis after removing strand invaders.

```{r voom-filtered, dev='png'}

# In alphabetical order
samples <- c(ddd_nw_2, lll_nw_2, rrr_nw_2)
groups  <- c(rep('DNA_nw2',3), rep('LNA_nw2',3), rep('RNA_nw2', 3))

design <- model.matrix(~0+groups)
colnames(design) <- c("DNA_nw2", "LNA_nw2", "RNA_nw2")
design

cont.matrix <- makeContrasts( LNA_vs_RNA = LNA_nw2 - RNA_nw2
                            , RNA_vs_DNA = RNA_nw2 - DNA_nw2
                            , DNA_vs_LNA = DNA_nw2 - LNA_nw2
                            , levels = design)

d <-  DGEList( counts = l2[samples]
             , group  = groups
             , remove.zeros = TRUE) %>%
        calcNormFactors %>%
        voom(design, plot = TRUE)

plotMDS(d)

d %<>% lmFit(design) %>%
  contrasts.fit(cont.matrix) %>%
  eBayes

d$genes <- genesymbols[rownames(d),]

topTable(d, n=30)
summary(decideTests(d))
decideTests(d) %>% vennDiagram
```

### LNA vs RNA (filtered)

```{r compare_LNA_RNA_f}
n <- 1
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="LNA vs RNA (filtered)")
topTreat(d,n)
```

### RNA vs DNA (filtered)

```{r compare_RNA_DNA_f}
n <- 2
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="RNA vs DNA (filtered)")
topTreat(d,n)
```


### DNA vs LNA (filtered)

```{r compare_DNA_LNA_f}
n <- 3
cont.matrix[,n]
plotMD( d, coef=n, status=decideTests(d)[,n] !=0
      , main="DNA vs LNA (filtered)")
topTreat(d,n)
```

## Session info.

```{r session_info, cache=FALSE}
sessionInfo()
```
